core/src/main/scala/org/http4s/blaze/pipeline/stages/TimeoutStageBase.scala

```diff
@@ -22,14 +24,21 @@ abstract class TimeoutStageBase[T](timeout: Duration, exec: TickWheelExecutor) e
 
   private val killswitch = new Runnable {
     override def run(): Unit = {
+      logger.debug(s"Timeout of $timeout triggered. Killing pipeline.")
       sendOutboundCommand(Disconnect)
       sendInboundCommand(Disconnected)
     }
   }
 
-  protected def setAndCancel(next: Cancellable): Unit = {
+  @tailrec
+  final def setAndCancel(next: Cancellable): Unit = {
     val prev = lastTimeout.getAndSet(next)
-    if (prev != null) prev.cancel()
+    if (prev == closedTag && next != closedTag) {
+      // woops... we submitted a new cancellation when we were closed!
```

Comment:
This event is not always avoidable. For example (happens in http4s), the server pipeline attempts a read then the client disconnects. A `Disconnected` signal shuts down the pipeline including the timeout stage, but the pending read will also complete and the timeout stage will attempt to reset.